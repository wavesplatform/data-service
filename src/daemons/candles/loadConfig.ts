import { memoizeWith, always } from 'ramda';
import {
  PostgresConfig,
  LoggerConfig,
  loadDefaultConfig,
} from '../../loadConfig';

export type CandlesConfig = PostgresConfig &
  LoggerConfig & {
    candlesUpdateInterval: number;
    candlesUpdateTimeout: number;
    candlesTruncateTable: boolean;
    candlesTableName: string;
  };

const load = (): CandlesConfig => ({
  ...loadDefaultConfig(),
  candlesUpdateInterval: process.env.CANDLES_UPDATE_INTERVAL
    ? parseInt(process.env.CANDLES_UPDATE_INTERVAL)
    : 2500,
  candlesUpdateTimeout: process.env.CANDLES_UPDATE_TIMEOUT
    ? parseInt(process.env.CANDLES_UPDATE_TIMEOUT)
    : 20000,
  candlesTruncateTable: process.env.RECALCULATE_ALL_CANDLES_ON_START
    ? process.env.RECALCULATE_ALL_CANDLES_ON_START == 'true'
    : false,
  candlesTableName: process.env.CANDLES_TABLE_NAME || 'candles',
});

export const loadConfig = memoizeWith(always('config'), load);
