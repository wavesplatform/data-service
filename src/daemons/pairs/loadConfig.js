const checkEnv = require('check-env');

const { memoizeWith, always } = require('ramda');

const loadConfig = () => {
  // assert all necessary env vars are set
  checkEnv(['PGHOST', 'PGDATABASE', 'PGUSER', 'PGPASSWORD']);

  return {
    postgresHost: process.env.PGHOST,
    postgresPort: parseInt(process.env.PGPORT) || 5432,
    postgresDatabase: process.env.PGDATABASE,
    postgresUser: process.env.PGUSER,
    postgresPassword: process.env.PGPASSWORD,
    postgresPoolSize: parseInt(process.env.PGPOOLSIZE) || 20,

    logLevel: process.env.LOG_LEVEL || 'info',

    pairsUpdateInterval: process.env.PAIRS_UPDATE_INTERVAL || 2500,
    pairsUpdateTimeout: process.env.PAIRS_UPDATE_TIMEOUT || 20000,
  };
};

module.exports = memoizeWith(always('config'), loadConfig);
