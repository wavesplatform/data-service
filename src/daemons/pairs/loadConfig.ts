import { memoizeWith, always } from 'ramda';
import {
  PostgresConfig,
  LoggerConfig,
  loadDefaultConfig,
} from '../../loadConfig';

export type PairsConfig = PostgresConfig &
  LoggerConfig & {
    pairsUpdateInterval: number;
    pairsUpdateTimeout: number;
  };

const load = (): PairsConfig => ({
  ...loadDefaultConfig(),
  pairsUpdateInterval: process.env.PAIRS_UPDATE_INTERVAL
    ? parseInt(process.env.PAIRS_UPDATE_INTERVAL)
    : 2500,
  pairsUpdateTimeout: process.env.PAIRS_UPDATE_TIMEOUT
    ? parseInt(process.env.PAIRS_UPDATE_TIMEOUT)
    : 20000,
});

export const loadConfig = memoizeWith(always('config'), load);
