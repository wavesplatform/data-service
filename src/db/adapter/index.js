// query and formatting
const sql = require('../sql');
const { formatPairs, batchQuery } = require('../utils');

// error handling
const { toDbError } = require('../../errorHandling');

// db adapter factory
const createDbAdapter = taskedDBDriver => {
  return {
    assets: assetIdArr =>
      taskedDBDriver
        .many(sql.assets, [assetIdArr])
        // .map(batchQuery((reqId, { id }) => reqId === id, assetIdArr))
        .mapRejected(toDbError({ request: 'assets' })),
    // compose(
    // batchQuery((reqId, { id }) => reqId === id, assetIdArr),
    // a => { console.log(a); return a},
    // bind(db.many)
    // )(sql.assets, [assetIdArr]),
    // volumes: pairs => db.many(sql.volumes, formatPairs(pairs)),
  };
};

module.exports = createDbAdapter;
