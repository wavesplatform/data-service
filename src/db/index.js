const createTaskedDriver = require('./driver');
const createAdapter = require('./adapter');

module.exports = options => {
  const taskedDbDriver = createTaskedDriver(options);
  const adapter = createAdapter(taskedDbDriver);
  return adapter;
};
