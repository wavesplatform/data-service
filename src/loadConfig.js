const checkEnv = require('check-env');

const { memoizeWith, always } = require('ramda');

const loadConfig = () => {
  // assert all necessary env vars are set
  checkEnv(['PGHOST', 'PGDATABASE', 'PGUSER', 'PGPASSWORD']);

  return {
    port: parseInt(process.env.PORT) || 3000,
    postgresHost: process.env.PGHOST,
    postgresPort: parseInt(process.env.PGPORT) || 5432,
    postgresDatabase: process.env.PGDATABASE,
    postgresUser: process.env.PGUSER,
    postgresPassword: process.env.PGPASSWORD,
    postgresPoolSize: parseInt(process.env.PGPOOLSIZE) || 20,

    redisPort: parseInt(process.env.REDIS_PORT) || undefined,
    redisHost: process.env.REDIS_HOST,
    redisPassword: process.env.REDIS_PASSWORD,

    logLevel: process.env.LOG_LEVEL || 'info',
  };
};

module.exports = memoizeWith(always('config'), loadConfig);
