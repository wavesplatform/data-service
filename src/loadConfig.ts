import * as checkEnv from 'check-env';
import { memoizeWith, always } from 'ramda';

export type PostgresConfig = {
  postgresHost: string;
  postgresPort: number;
  postgresDatabase: string;
  postgresUser: string;
  postgresPassword: string;
  postgresPoolSize: number;
};

export type LoggerConfig = {
  logLevel: string;
};

export type ServerConfig = {
  port: number;
};

export type MatcherConfig = {
  matcher: {
    settingsURL?: string;
    defaultMatcherAddress: string;
  };
};

export type DefaultConfig = PostgresConfig & ServerConfig & LoggerConfig;

export type DataServiceConfig = PostgresConfig &
  ServerConfig &
  LoggerConfig &
  MatcherConfig & {
    defaultTimeout: number;
  };

const commonEnvVariables = ['PGHOST', 'PGDATABASE', 'PGUSER', 'PGPASSWORD'];

export const loadDefaultConfig = (): DefaultConfig => {
  // assert common env vars are set
  checkEnv(commonEnvVariables);

  return {
    port: process.env.PORT ? parseInt(process.env.PORT) : 3000,
    postgresHost: process.env.PGHOST || '',
    postgresPort: process.env.PGPORT ? parseInt(process.env.PGPORT) : 5432,
    postgresDatabase: process.env.PGDATABASE || 'mainnet',
    postgresUser: process.env.PGUSER || 'postgres',
    postgresPassword: process.env.PGPASSWORD || 'postgres',
    postgresPoolSize: process.env.PGPOOLSIZE
      ? parseInt(process.env.PGPOOLSIZE)
      : 20,
    logLevel: process.env.LOG_LEVEL || 'info',
  };
};

const envVariables = ['DEFAULT_MATCHER'];

const load = (): DataServiceConfig => {
  // assert all necessary env vars are set
  checkEnv(envVariables);

  const matcher: MatcherConfig = {
    matcher: {
      defaultMatcherAddress: process.env.DEFAULT_MATCHER as string,
    },
  };

  if (
    typeof process.env.MATCHER_SETTINGS_URL !== 'undefined' &&
    process.env.MATCHER_SETTINGS_URL !== ''
  ) {
    matcher.matcher.settingsURL = process.env.MATCHER_SETTINGS_URL;
  }

  return {
    ...loadDefaultConfig(),
    ...matcher,

    defaultTimeout: process.env.DEFAULT_TIMEOUT
      ? parseInt(process.env.DEFAULT_TIMEOUT)
      : 30000,
  };
};

export const loadConfig = memoizeWith(always('config'), load);
