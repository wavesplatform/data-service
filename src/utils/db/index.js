const Maybe = require('folktale/maybe');
const { compose, map, filter, has, __, curryN, find } = require('ramda');

/**
 * matchFn :: (Request, Result) -> Boolean
 *
 * matchRequestsResults ::
 *  matchFn ->
 *  Request[] ->
 *  Result[] ->
 *  (Maybe Result)[]
 * */
const matchRequestsResults = curryN(3, (matchFn, requests, results) => {
  const findResult = compose(
    Maybe.fromNullable,
    req => find(res => matchFn(req, res), results)
  );
  return map(findResult, requests);
});

/**
 * filter :: Query -> QueryWithFilter
 *
 * pickBindFilters ::
 *    { [fName]: fValue -> filter } ->
 *    fName[] ->
 *    { [fName]: fValue } ->
 *    filter[]
 * */
const pickBindFilters = curryN(3, (F, fsToApply, fValues) =>
  compose(
    map(x => F[x](fValues[x])),
    filter(has(__, fValues))
  )(fsToApply)
);

/**
 * Escapes query for SQL condition with to_tsquery function
 * @param {string} query
 * @returns {string}
 */
const escapeForTsQuery = query => {
  return query
    .replace(/[^\w\s]|_/g, '')
    .trim()
    .replace(/\s+/g, ' & ')
};

/**
 * Escapes query for SQL condition with like operator
 * @param {string} query
 * @returns {string}
 */
const escapeForLike = query => query.replace(/%/g, '\\%');

/**
 * Prepares query for SQL condition with like operator
 * @param {string} query
 * @param {object} params
 * @returns {string}
 */
const prepareForLike = (query, params = { matchExactly: false }) =>
  compose(
    query => (params.matchExactly ? query : `${query}%`),
    escapeForLike
  )(query);

module.exports = {
  matchRequestsResults,
  pickBindFilters,
  escapeForTsQuery,
  escapeForLike,
  prepareForLike,
};
