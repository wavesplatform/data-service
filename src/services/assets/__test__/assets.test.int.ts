const http = require('http');
const json = require('@waves/json-bigint');

import createService, { createCache } from '../index';

// dependencies
import { createPgDriver } from '../../../db';
import { loadConfig } from '../../../loadConfig';
import { EventEmitter } from 'events';

const options = loadConfig();
const drivers = {
  pg: createPgDriver(options),
};
const cache = createCache(10, 10000);

const DEFAULT_TIMEOUT_IN_MS = 30000;

const service = createService({
  drivers,
  emitEvent: () => () => null,
  cache,
  timeouts: {
    get: DEFAULT_TIMEOUT_IN_MS,
    mget: DEFAULT_TIMEOUT_IN_MS,
    search: DEFAULT_TIMEOUT_IN_MS,
  },
});

const assetId = 'G8VbM7B6Zu8cYMwpfRsaoKvuLVsy8p1kYP4VvSdwxWfH';

describe('Assets service', () => {
  describe('get', () => {
    it('fetches a real asset', async done => {
      service
        .get(assetId)
        .run()
        .promise()
        .then(x => {
          expect(x.unsafeGet()).toMatchSnapshot();
          done();
        })
        .catch(done.fail);
    });

    it('returns null for unreal tx', async () => {
      const tx = await service
        .get('UNREAL')
        .run()
        .promise();

      expect(tx).toBeNothing();
    });
  });

  describe('mget', () => {
    it('fetches real assets with nulls for unreal', async done => {
      service
        .mget([assetId, 'UNREAL'])
        .run()
        .promise()
        .then(xs => {
          expect(xs).toMatchSnapshot();
          done();
        })
        .catch(e => done(JSON.stringify(e)));
    });
  });

  describe('search', () => {
    it('fetches WAVES by ticker', async done => {
      service
        .search({ ticker: 'WAVES', limit: 100 })
        .run()
        .promise()
        .then(xs => {
          expect(xs).toMatchSnapshot();
          done();
        })
        .catch(e => done(JSON.stringify(e)));
    });

    it('fetches non-WAVES asset by ticker (BTC)', async done => {
      http.get(
        'http://nodes.wavesnodes.com/assets/details/8LQW8f7P5d5PZM7GtZEBgaqRPGSzS3DfPuiXrURJ4AJS',
        (res: EventEmitter) => {
          let data: string = '';
          res.on('data', (chunk: string) => (data += chunk));
          res.on('end', () => {
            const assetInfoFromNode: any = json.parse(data);
            service
              .search({ ticker: 'BTC', limit: 100 })
              .run()
              .promise()
              .then(xs => {
                expect(xs.data[0].data).toMatchObject({
                  description: assetInfoFromNode.description,
                  height: assetInfoFromNode.issueHeight,
                  id: assetInfoFromNode.assetId,
                  name: assetInfoFromNode.name,
                  precision: assetInfoFromNode.decimals,
                  quantity: assetInfoFromNode.quantity,
                  reissuable: assetInfoFromNode.reissuable,
                  sender: assetInfoFromNode.issuer,
                  ticker: 'BTC',
                });
                done();
              });
          });
        }
      );
    });

    it('fetches all assets with tickers by ticker=*', () =>
      service
        .search({ ticker: '*', limit: 100 })
        .run()
        .promise()
        .then(as => {
          expect(as.data.length).toBeGreaterThan(100);
          // make sure WAVES is included
          expect(
            as.data.find(a => a.data && a.data.ticker === 'WAVES')
          ).not.toBeUndefined();
        }));
  });
});
