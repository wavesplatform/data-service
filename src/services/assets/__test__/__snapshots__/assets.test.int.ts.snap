// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assets service get fetches a real asset 1`] = `
Object {
  "__type": "asset",
  "data": Object {
    "description": "ETLxxx",
    "hasScript": false,
    "height": 238692,
    "id": "G8VbM7B6Zu8cYMwpfRsaoKvuLVsy8p1kYP4VvSdwxWfH",
    "minSponsoredFee": null,
    "name": "ETLxxx",
    "precision": 8,
    "quantity": "100000000",
    "reissuable": false,
    "sender": "3P7HKYe6HB8eK7uXxrztEYUSzZQyyYRnWHg",
    "ticker": null,
    "timestamp": 2016-11-19T19:45:07.045Z,
  },
}
`;

exports[`Assets service mget fetches real assets with nulls for unreal 1`] = `
Object {
  "__type": "list",
  "data": Array [
    Object {
      "__type": "asset",
      "data": Object {
        "description": "ETLxxx",
        "hasScript": false,
        "height": 238692,
        "id": "G8VbM7B6Zu8cYMwpfRsaoKvuLVsy8p1kYP4VvSdwxWfH",
        "minSponsoredFee": null,
        "name": "ETLxxx",
        "precision": 8,
        "quantity": "100000000",
        "reissuable": false,
        "sender": "3P7HKYe6HB8eK7uXxrztEYUSzZQyyYRnWHg",
        "ticker": null,
        "timestamp": 2016-11-19T19:45:07.045Z,
      },
    },
    Object {
      "__type": "asset",
      "data": null,
    },
  ],
}
`;

exports[`Assets service search fetches WAVES by ticker 1`] = `
Object {
  "__type": "list",
  "data": Array [
    Object {
      "__type": "asset",
      "data": Object {
        "description": "",
        "hasScript": false,
        "height": 0,
        "id": "WAVES",
        "minSponsoredFee": null,
        "name": "Waves",
        "precision": 8,
        "quantity": "10000000000000000",
        "reissuable": false,
        "sender": "",
        "ticker": "WAVES",
        "timestamp": 2016-04-11T18:00:00.000Z,
      },
    },
  ],
}
`;
